Add roscontrol

serraramiro1 opened the pull request:
# Description

Adds roscontrol gazebo plugin, as well as a controller, and a hardware interface node to interface with the real robot

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

```bash
# Step 1
RVIZ=True RVIZ_CONFIG=navigation LOCALIZATION=amcl ROS_CONTROL=True roslaunch ca_gazebo create_house.launch 
# Step 2
use rviz UI to send goals
```
with real robot
```bash
# Step 1
ROS_CONTROL=true roslaunch ca_bringup minimal.launch 
# Step 2
roslaunch ca_tools keyboard_teleop.launch
```

**Test Configuration**:

- [x] Real robot
- [x] Gazebo simulation
- [ ] Other (please specify)

# Checklist:

- [x] My code follows the style guidelines of this project (Travis CI is passing)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

eborghi10 commented:
Travis is failing, please check [this](https://travis-ci.com/github/RoboticaUtnFrba/create_autonomy/builds/171868737#L8632) to fix it.
The command that it uses is this and you try it locally:

```
catkin lint -W2 --strict --explain $CREATE_AUTONOMY_SRC --ignore plugin_missing_install --ignore target_name_collision
```

Replace `$CREATE_AUTONOMY_SRC` with the path to `create_autonomy`.

eborghi10 commented:
Would you update the description of the PR?

eborghi10 commented on ca_description/launch/create_description.launch:
```diff
@@ -7,6 +7,7 @@
   <arg name="model" value="$(find ca_description)/urdf/create_2.xacro" doc="Path to Xacro description of the robot"/>
 
   <arg name="visualize" default="$(optenv VISUALIZE false)" doc="Flag to visualize rays in Gazebo"/>
+  <arg name="enable_roscontrol" default="$(optenv ROSCONTROL False)" doc="Flag to use roscontrol Gazebo plugin"/>
```
Use `ROS_CONTROL` and put it in `true` by default.

```suggestion
  <arg name="enable_ros_control" default="$(optenv ROS_CONTROL true)" doc="Flag to use ros_control."/>
```

I also suggested a modification in `doc` because this flag can also be used with the real robot.

serraramiro1 replied:
Mmm, I don't agree with this...
The cmd_vel topic that this exposes doesn't quite match with the existing one, and can't be remapped.
Making move_base work with this will require a little extra work

eborghi10 replied:
Ok, but address the rest of the sentence.

What's the difference?

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,38 @@
-        <odometrySource>world</odometrySource>
-        <publishTf>true</publishTf>
-      </plugin>
+      <xacro:if value="${not $(arg enable_roscontrol)}">
```
```suggestion
      <xacro:unless value="$(arg enable_ros_control)">
```

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,38 @@
+      <!-- ros-control plugin -->
+      <xacro:if value="$(arg enable_roscontrol)">
+      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
+        <robotNamespace>/create1</robotNamespace>
```
You have more parameters to parametrize the plugin: https://github.com/ros-simulation/gazebo_ros_pkgs/blob/melodic-devel/gazebo_ros_control/src/gazebo_ros_control_plugin.cpp#L84-L159

eborghi10 commented on navigation/create_control/CMakeLists.txt:
```diff
@@ -0,0 +1,204 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(create_control)
+
+## Compile as C++11, supported in ROS Kinetic and newer
```
Remove commented code.

eborghi10 commented on navigation/create_control/CMakeLists.txt:
```diff
@@ -0,0 +1,204 @@
+## DEPENDS: system dependencies of this project that dependent projects also need
+catkin_package(
+  LIBRARIES create_control
+  CATKIN_DEPENDS controller_manager joint_state_controller robot_state_publisher libcreate
```
Put one dependency on each line.

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,75 @@
+#include <hardware_interface/robot_hw.h>
+#include <ros/ros.h>
+
+class Create2_interface : public hardware_interface::RobotHW {
```
Classes have to be `CamelCased`.

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,75 @@
+    model_ = create::RobotModel::CREATE_2;
+    robot_.reset(new create::Create(model_));
+
+    if (!robot_->connect("/dev/ttyUSB0",
```
You can get these parameters from [here](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_node/launch/create_2.launch#L10).

eborghi10 replied:
Load the parameters from the YAML file.

serraramiro1 replied:
This is still pending

eborghi10 replied:
I'll wait for this change to test it again.

serraramiro1 replied:
Addressed

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,13 @@
+
+  <!-- Load controller configuration from YAML file to parameter server -->
+  <arg name="real_robot" value="$(optenv REAL_ROBOT false)"/>
+  <node name="robot_hw_interface" pkg="create_control" type="create_control_node" if="$(arg real_robot)"
```
I don't have a clear solution in mind but I would try to avoid `REAL_ROBOT`. Maybe since you need to call `minimal.launch`, this launch file can set it directly.

serraramiro1 replied:
Still pending, I'll try to fix real robot stuff if I get to test it on one...

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,13 @@
+  <!-- Load controller configuration from YAML file to parameter server -->
+  <arg name="real_robot" value="$(optenv REAL_ROBOT false)"/>
+  <node name="robot_hw_interface" pkg="create_control" type="create_control_node" if="$(arg real_robot)"
+    ns="/create1"/>
```
Don't hard-code the namespace.

eborghi10 commented on navigation/create_control/package.xml:
```diff
@@ -0,0 +1,27 @@
+<package format="2">
+  <name>create_control</name>
+  <version>0.0.0</version>
+  <description>The create_control package</description>
```
Sorry but Travis will complain about this description :sweat_smile:

eborghi10 commented on ca_description/launch/create_description.launch:
```diff
@@ -7,6 +7,7 @@
   <arg name="model" value="$(find ca_description)/urdf/create_2.xacro" doc="Path to Xacro description of the robot"/>
 
   <arg name="visualize" default="$(optenv VISUALIZE false)" doc="Flag to visualize rays in Gazebo"/>
+  <arg name="enable_ros_control" default="$(optenv ROS_CONTROL False)" doc="Flag to use roscontrol Gazebo plugin"/>
```
```suggestion
  <arg name="enable_ros_control" default="$(optenv ROS_CONTROL False)" doc="Flag to use ros_control Gazebo plugin"/>
```

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,39 @@
   <xacro:macro name="sim_create_base">
 
     <gazebo>
+      <xacro:arg name="enable_roscontrol" default="False"/>
```
```suggestion
      <xacro:arg name="enable_ros_control" default="False"/>
```

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,39 @@
-        <odometrySource>world</odometrySource>
-        <publishTf>true</publishTf>
-      </plugin>
+      <xacro:unless value="$(arg enable_roscontrol)">
```
```suggestion
      <xacro:unless value="$(arg enable_ros_control)">
```

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,39 @@
+        </plugin>
+    </xacro:unless>
 
+      <!-- ros-control plugin -->
+      <xacro:if value="$(arg enable_roscontrol)">
```
```suggestion
      <!-- ros_control plugin -->
      <xacro:if value="$(arg enable_ros_control)">
```

eborghi10 commented on ca_description/urdf/create_base_gazebo.xacro:
```diff
@@ -3,28 +3,39 @@
+      <!-- ros-control plugin -->
+      <xacro:if value="$(arg enable_roscontrol)">
+      <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
+        <robotNamespace>/create1</robotNamespace>
```
This namespace can change
```suggestion
        <robotNamespace>${robot_name}</robotNamespace>
```

eborghi10 commented on ca_gazebo/launch/create_empty_world.launch:
```diff
@@ -7,6 +7,7 @@
 	<arg name="paused"  value="$(optenv PAUSED false)"  doc="Start the simulation paused"/>
 	<arg name="debug"   value="$(optenv DEBUG false)" 	doc="Enable debugging with gdb"/>
 	<arg name="verbose" value="$(optenv VERBOSE false)" doc="Enable verbose output in Gazebo"/>
+	<arg name="enable_ros_control" value="$(optenv ROS_CONTROL false)" doc="Enable enable_ros_control"/>
```
```suggestion
	<arg name="enable_ros_control" value="$(optenv ROS_CONTROL false)" doc="Flag to use ros_control Gazebo plugin"/>
```

eborghi10 commented on navigation/create_control/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+# Mark cpp header files for installation
+install(DIRECTORY include/
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+  FILES_MATCHING PATTERN "*.h"
+  PATTERN ".svn" EXCLUDE
```
```suggestion
```

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,75 @@
+  ros::Time last_update_;
+  std::unique_ptr<create::Create> robot_;
+  create::RobotModel model_;
+};
```
```suggestion
};

```

eborghi10 commented on navigation/create_control/src/create_control/create_control_node.cpp:
```diff
@@ -0,0 +1,17 @@
+    robot.write();
+    r.sleep();
+  }
+}
```
```suggestion
}

```

eborghi10 commented on docker/create_ros_melodic_gazebo9/packages.txt:
```diff
@@ -14,6 +14,7 @@ ros-*-cartographer-ros-msgs
 ros-*-cartographer-rviz
 ros-*-controller-manager
 ros-*-depthimage-to-laserscan
+ros-*-diff-drive-controller
```
But put it in alphabetical order:

```suggestion
ros-*-diff-drive-controller
ros-*-joint-state-controller
```

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,18 @@
+
+  <!-- load the controller -->
+  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
+    output="screen" ns="$(arg ns)" args="mobile_base_controller"/>
```
You don't need `ns` here if you move `create_control.launch` inside `spawn_create.launch`.

```suggestion
    output="screen" args="mobile_base_controller"/>
```

eborghi10 commented:
Gazebo (`ros_control` plugin?) is publishing the relationship `odom` --> `base_link` without the tf_prefix (`createN_tf`):

![Screenshot from 2020-06-26 20-22-46](https://user-images.githubusercontent.com/3400230/85908287-e4f96000-b7ea-11ea-9f00-b86b4ba1153e.png)

Check the parameters here to solve this issue: http://wiki.ros.org/diff_drive_controller

Also put a link to this wiki in `create_2_control.yaml` and complete the missing parameters.

eborghi10 commented on ca_gazebo/launch/create_empty_world.launch:
```diff
@@ -32,5 +33,7 @@
     <arg name="env"          value="$(arg env)"/>
 		<arg name="localization" value="$(arg localization)" />
 	</include>
+	<include file="$(find create_control)/launch/create_control.launch" if="$(arg enable_ros_control)"/>
```
Since this is unique per robot, it should be moved to `spawn_create.launch`.

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,18 @@
+    ns="$(arg ns)">
+    <rosparam command="load" file="$(find ca_driver)/config/default.yaml"/>
+  </node>
+  <rosparam file="$(find create_control)/config/create_2_control.yaml" command="load"/>
```
Should this be moved inside a node?

serraramiro1 replied:
Doesn't work if so

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,7 @@
+create1:
+  mobile_base_controller:
+    type: "diff_drive_controller/DiffDriveController"
+    left_wheel: 'wheel_right_joint'
+    right_wheel: 'wheel_left_joint'
+    pose_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
+    twist_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
```
`create1` is too specific for one robot. And also is not needed.

```suggestion
mobile_base_controller:
  type: "diff_drive_controller/DiffDriveController"
  left_wheel: 'wheel_right_joint'
  right_wheel: 'wheel_left_joint'
  pose_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
  twist_covariance_diagonal: [0.001, 0.001, 1000000.0, 1000000.0, 1000000.0, 1000.0]
```

You can easily debug it running this:

```
GUI=false RVIZ=True ROS_CONTROL=True roslaunch ca_gazebo create_house.launch
```

serraramiro1 commented:
Addressed most of the comments, please re-review @eborghi10

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,43 @@
+  # diff_drive_controller will attempt to read either one or both from the
+  # URDF if not specified as a parameter
+  wheel_separation : 1.0
+  wheel_radius : 0.3
```
I took the values from [here](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_description/urdf/create_2.xacro#L8-L11).

```suggestion
  wheel_separation : 0.235
  wheel_radius : 0.036
```

serraramiro1 replied:
Changed

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,43 @@
+  wheel_radius : 0.3
+
+  # Wheel separation and radius multipliers
+  wheel_separation_multiplier: 1.0 # default: 1.0
+  wheel_radius_multiplier    : 1.0 # default: 1.0
```
Interesting, these parameters are used in the real robot to do a wheel calibration. This means that if you want to move the robot straight, it won't do it because the wheel radius and separation are not exactly what the parameters above say.

There is a test called UMBmark that allows you to find these values.

serraramiro1 replied:
Could you give me a pointer where to find it?

eborghi10 replied:
You don't need to implement it in case you have access to a real robot, but in [this issue](https://github.com/RoboticaUtnFrba/create_autonomy/issues/57) you can find some material. And particularly for UMBmark, see [this](http://thetechnicgear.com/2014/06/howto-calibrate-differential-drive-robot/).

eborghi10 replied:
Can you make a comment about what I told you here?

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,43 @@
+  wheel_radius_multiplier    : 1.0 # default: 1.0
+
+  # Velocity commands timeout [s], default 0.5
+  cmd_vel_timeout: 0.25
```
Was this value set based on testing? Keep in mind that [here](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_driver/config/default.yaml#L16-L17) we use `0.2` instead.

serraramiro1 replied:
Set to 0.2

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,43 @@
+  # Velocity commands timeout [s], default 0.5
+  cmd_vel_timeout: 0.25
+
+  # Velocity and acceleration limits
```
I can't find where are these params in the real robot with the other approach. It would be interesting to understand how the code deals with these parameters because there is a node (that is not set up already) but you can do something similar (safety_limit).

serraramiro1 replied:
https://github.com/ros-controls/ros_controllers/blob/noetic-devel/diff_drive_controller/src/speed_limiter.cpp

It basically clamps them, I assume safety_limit does the same

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,76 @@
+#pragma once
```
The idea behind this node is to replace [this node](https://github.com/RoboticaUtnFrba/create_autonomy/blob/kinetic-devel/ca_driver/launch/create_2.launch#L7-L10) but if you enable it now, all the other sensors won't be available.

To don't forget to do it, I recommend you to create a README at the level of `create_control` and mention this and other technical debts that we may find.

serraramiro1 replied:
Added

serraramiro1 commented:
> I made a few comments but I'll give it a try in sim & real robot later.

I'm pretty sure this won't work as is, I remember having to change the spinner in the hw interface to an async spinner,
I can't test it for obvious reasons

eborghi10 commented:
> > I made a few comments but I'll give it a try in sim & real robot later.
> 
> I'm pretty sure this won't work as is, I remember having to change the spinner in the hw interface to an async spinner,
> I can't test it for obvious reasons

I can give you access to the robot so you can debug the code.

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,43 @@
+  type        : "diff_drive_controller/DiffDriveController"
+  left_wheel  : 'wheel_left_joint'
+  right_wheel : 'wheel_right_joint'
+  publish_rate: 50.0               # default: 50
```
Gazebo and real robot run at 30 Hz.

```suggestion
  publish_rate: 30.0               # default: 50
```

serraramiro1 replied:
done

eborghi10 commented:
Can you rebase with latest `kinetic-devel`?

serraramiro1 commented:
> Can you rebase with latest `kinetic-devel`?

done

eborghi10 commented:
> ```
> [ERROR] [1593214361.220911060, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_left_joint                              │
> [ERROR] [1593214361.221558857, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_right_joint  
> ```

Can you check if this can be fixed? Hint is [here](https://answers.gazebosim.org/question/5468/gazebo_ros_pkgs-no-p-gain-specified-for-pid-in-positionjointinterface/?answer=5470#post-id-5470).

eborghi10 commented:
`move_base` is not working. You can check two things to debug it:

1. Topic remappings with `rqt_graph`. I checked that `odom` is not connected between `move_base` and `ros_control`. Also the velocity command is unconnected.
2. Check the Tf prefixes.

serraramiro1 commented:
> > ```
> > [ERROR] [1593214361.220911060, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_left_joint                              │
> > [ERROR] [1593214361.221558857, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_right_joint  
> > ```
> 
> Can you check if this can be fixed? Hint is [here](https://answers.gazebosim.org/question/5468/gazebo_ros_pkgs-no-p-gain-specified-for-pid-in-positionjointinterface/?answer=5470#post-id-5470).

Set to 1 0 0, changing `d` value ends in odd results...

serraramiro1 commented:
> `move_base` is not working. You can check two things to debug it:
> 
> 1. Topic remappings with `rqt_graph`. I checked that `odom` is not connected between `move_base` and `ros_control`. Also the velocity command is unconnected.
> 2. Check the Tf prefixes.

Addressed, added an external republisher, since controller's topics cant be remapped

eborghi10 commented:
> > > ```
> > > [ERROR] [1593214361.220911060, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_left_joint                              │
> > > [ERROR] [1593214361.221558857, 0.956000000]: No p gain specified for pid.  Namespace: /create1/gazebo_ros_control/pid_gains/wheel_right_joint  
> > > ```
> > 
> > 
> > Can you check if this can be fixed? Hint is [here](https://answers.gazebosim.org/question/5468/gazebo_ros_pkgs-no-p-gain-specified-for-pid-in-positionjointinterface/?answer=5470#post-id-5470).
> 
> Set to 1 0 0, changing `d` value ends in odd results...

I suspect that you should add at least 0.1 to the integral part (the second one).

eborghi10 commented:
> > `move_base` is not working. You can check two things to debug it:
> > 
> > 1. Topic remappings with `rqt_graph`. I checked that `odom` is not connected between `move_base` and `ros_control`. Also the velocity command is unconnected.
> > 2. Check the Tf prefixes.
> 
> Addressed, added an external republisher, since controller's topics cant be remapped

Have you checked if the code doesn't have a way to changing it? Maybe it would be better to have our own copy of the code.

eborghi10 commented on ca_bringup/launch/minimal_ros_control.launch:
```diff
@@ -0,0 +1,20 @@
+  </node>
+
+  <!-- Ros control -->
+  <include file="$(find create_control)/launch/create_control.launch" ns="$(arg ns)"/>
```
Taking into account that `ROS_CONTROL` is being used for the simulation, it won't be a bad idea to do the same in `minimal.launch`.

serraramiro1 replied:
Changed to use ROS_CONTROL roslaunch ca_bringup minimal.launch

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,20 @@
+<launch>
+  <arg name="robot_id"  default="$(optenv ID 1)"      doc="Unique identifier of the robot [1-Inf.)"/>
+  <arg name="ns" value="create$(arg robot_id)"/>
```
```suggestion
  <arg name="ns" value="create$(arg robot_id)"/>
  <arg name="tf_prefix" value="$(arg ns)_tf"          doc="Tf prefix"/>
```

serraramiro1 replied:
Done

eborghi10 commented on navigation/create_control/launch/create_control.launch:
```diff
@@ -0,0 +1,20 @@
+
+  <!-- need to be set separatedly, since they depend on the robot id, and that cannot be parametrized
+  in the yaml file.-->
+  <param name="/create1/mobile_base_controller/odom_frame_id" value="/create$(arg robot_id)_tf/odom"/>
+  <param name="/create1/mobile_base_controller/base_frame_id" value="/create$(arg robot_id)_tf/base_footprint"/>
```
I don't think you will need a namespace in the parameter because the whole launch file is already there.

```suggestion
  <param name="mobile_base_controller/odom_frame_id" value="$(arg tf_prefix)/odom"/>
  <param name="mobile_base_controller/base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
```

serraramiro1 replied:
Done

eborghi10 commented on navigation/create_control/scripts/ros_control_republishers.py:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/python
```
From [this issue](https://github.com/ros-controls/ros_controllers/issues/126), you can use this:

http://wiki.ros.org/topic_tools/relay

eborghi10 replied:
Also, keep in mind that topic names shouldn't be hardcoded.

eborghi10 replied:
You should add this package to `packages.txt`.

serraramiro1 replied:
Done

eborghi10 commented on navigation/create_control/src/create_control/create_control_node.cpp:
```diff
@@ -0,0 +1,19 @@
+  Create2Interface robot(nh);
+  controller_manager::ControllerManager cm(&robot, nh);
+
+  ros::Rate r(100);
```
Create a node parameter to change this rate.

serraramiro1 replied:
done

eborghi10 commented on navigation/create_control/src/create_control/create_control_node.cpp:
```diff
@@ -0,0 +1,19 @@
+    robot.write();
+    r.sleep();
+  }
+}
```
```suggestion
}

```

serraramiro1 replied:
Done

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,98 @@
+  void write() {
+    robot_->driveWheels(cmd[0] * (model_.getWheelDiameter()),
+                        cmd[1] * (model_.getWheelDiameter()));
+    // publishJointState();
```
Remove if it's not used.

```suggestion
```

serraramiro1 replied:
Removed

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,98 @@
+      ros::shutdown();
+    }
+    ROS_INFO("[CREATE] Connection established");
+    // Start in full control mode
+
+    robot_->setMode(create::MODE_FULL);
```
```suggestion

        // Start in full control mode
    robot_->setMode(create::MODE_FULL);
```

serraramiro1 replied:
Addressed

eborghi10 commented on navigation/create_control/include/create_hw_interface.hpp:
```diff
@@ -0,0 +1,98 @@
+    priv_nh_.param<std::string>("tf_prefix", tf_prefix_, "create1_tf");
+
+    robot_.reset(new create::Create(model_));
+    ROS_INFO_STREAM("\n\n" << dev << baud);
```
Use a proper log indication. You can take it from the other node.

serraramiro1 replied:
Deleted, this is not needed, and was only added for debugging purposes

eborghi10 commented on navigation/create_control/config/create_2_control.yaml:
```diff
@@ -0,0 +1,48 @@
+  linear:
+    x:
+      has_velocity_limits    : true
+      max_velocity           : 0.3  # m/s
+      min_velocity           : -0.2 # m/s
```
Aren't these too slow?

```suggestion
      max_velocity           : 0.5  # m/s
      min_velocity           : -0.4  # m/s
```

eborghi10 replied:
Same for other limits.

serraramiro1 replied:
Done

eborghi10 commented on ca_bringup/launch/minimal_ros_control.launch:
```diff
@@ -0,0 +1,20 @@
+  </include>
+
+  <!-- Load robot's hw interface -->
+  <node name="robot_hw_interface" pkg="create_control" type="create_control_node" ns="$(arg ns)"
```
Shouldn't this be moved inside `create_control`? Maybe it will affect the simulation so you can do this:

```
# File: create_control_hardware.launch
--> Run robot_hw_interface
--> Call create_control.launch
```

serraramiro1 replied:
Done

eborghi10 commented:
I'm seeing this issue when running with the real robot:

```
Loading controller: mobile_base_controller
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/controller_manager/spawner", line 212, in <module>
    if __name__ == '__main__': main()
  File "/opt/ros/melodic/lib/controller_manager/spawner", line 190, in main
    resp = load_controller(name)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 442, in __call__
    return self.call(*args, **kwds)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 512, in call
    raise ServiceException("unable to connect to service: %s"%e)
rospy.service.ServiceException: unable to connect to service: [Errno 104] Connection reset by peer
[INFO] [1593792112.905556]: Shutting down spawner. Stopping and unloading controllers...
[INFO] [1593792112.907553]: Stopping all controllers...
[WARN] [1593792112.911548]: Controller Spawner error while taking down controllers: unable to connect to service: [Errno 111] Connection refused
[create1/robot_hw_interface-4] process has died [pid 51963, exit code -6, cmd /create_ws/devel/lib/create_control/create_control_node __name:=robot_hw_interface __log:=/home/create/.ros/log/7f7d1f76-bd46-11ea-a694-9cb6d0e33077/create1-robot_hw_interface-4.log].
log file: /home/create/.ros/log/7f7d1f76-bd46-11ea-a694-9cb6d0e33077/create1-robot_hw_interface-4*.log
[create1/controller_spawner-5] process has died [pid 51969, exit code 1, cmd /opt/ros/melodic/lib/controller_manager/spawner mobile_base_controller __name:=controller_spawner __log:=/home/create/.ros/log/7f7d1f76-bd46-11ea-a694-9cb6d0e33077/create1-controller_spawner-5.log].
log file: /home/create/.ros/log/7f7d1f76-bd46-11ea-a694-9cb6d0e33077/create1-controller_spawner-5*.log
```
